lib/Data/OpenSRS.hs

```diff
@@ -115,50 +120,59 @@ doRequest' parser r = do
 
 -- | Transforms a SRSResponse into an error string
 responseError :: SRSResponse -> String
-responseError resp = show (srsResponseCode resp) ++ ": " ++ srsResponseText resp
+responseError resp = concat [show $ srsResponseCode resp,
+                             ": ",
+                             srsResponseText resp]
 
 -- | Posts request to OpenSRS
 postRequest :: SRSRequest -> IO (Response BSL8.ByteString)
-postRequest req =
+postRequest req = do
+    void $ tryDebug (requestConfig req) (BSL8.unpack . toLazyByteString . render . requestXML $ req)
     postWith opts (srsEndpoint $ requestConfig req) postBody
   where
     opts = defaults
-            & header "X-Username" .~ [pack $ srsUsername $ requestConfig req]
+            & header "X-Username" .~ [pack . srsUsername $ requestConfig req]
             & header "X-Signature" .~ [pack sig]
-            & header "Content-Length" .~ [pack $ show $ BSL8.length ps]
+            & header "Content-Length" .~ [pack . show $ BSL8.length ps]
     sig = md5Wrap (srsKey $ requestConfig req) (BSL8.unpack ps)
     postBody = XmlPost ps
-    ps = toLazyByteString $ render $ requestXML req
+    ps = toLazyByteString . render $ requestXML req
 
 -- | MD5 signature generator for OpenSRS
 md5Wrap :: String -> String -> String
-md5Wrap pk content = md5pack (md5pack (content ++ pk) ++ pk)
+md5Wrap pk content = md5pack (md5pack (content <> pk) <> pk)
   where
     md5pack = md5s . Str
 
+-- | Debug an XML string.
+tryDebug :: SRSConfig -> String -> IO ()
+tryDebug cfg s = when (srsDebug cfg) . traceIO $ prettyXML "  " s
+
+--------------------------------------------------------------------------------
+-- | Internal XML helper: Turns XML string into a list of XML TagTrees
+toXmlTree :: String -> [TagTree String]
+toXmlTree = tagTree . parseTags
+
 --------------------------------------------------------------------------------
 -- | Parse SRSResponse so we know if it's good or not
 parseResponse :: String -> SRSResponse
-parseResponse s = SRSResponse
-    (gt "<item key='is_success'>" == "1")
-    (gt "<item key='response_text'>")
-    (read $ gt "<item key='response_code'>")
-  where
+parseResponse s = let
     xml = parseTags s
-    gt = getText xml
+    in SRSResponse (getText xml "<item key='is_success'>" == "1")
+                   (getText xml "<item key='response_text'>")
+                   (read $ getText xml "<item key='response_code'>")
 
 --------------------------------------------------------------------------------
 -- | Parse domain list
 parseDomainList :: String -> DomainList
-parseDomainList s = DomainList (read $ gt "<item key='total'>")
-                               (Prelude.map makeDomain domainItems)
-                               (gt "<item key='remainder'>" == "1")
+parseDomainList s = DomainList (read $ getText xml "<item key='total'>")
```

Comment:
Read is never a good idea, not sure if it's 100% safe here, but read in production code is almost always bad.

As long as you have base >= 4.6, try to use readMaybe